qubit-sdk/src/main/java/com/qubit/android/sdk/internal/placement/interactor/PlacementInteractorImpl.kt

```diff
@@ -38,24 +41,54 @@ internal class PlacementInteractorImpl(
       onSuccess: OnPlacementSuccess,
       onError: OnPlacementError
   ) {
+    val placementMode = mode ?: DEFAULT_PLACEMENT_MODE
+    val attributes = buildAttributesJson(deviceId)
+    val cacheKey = buildCacheKey(placementId, placementMode, previewOptions, attributes)
     placementConnector.getPlacementModel(
         getPlacementApiHost(configurationRepository),
         placementId,
-        mode ?: DEFAULT_PLACEMENT_MODE,
-        deviceId,
+        placementMode,
         previewOptions,
-        { handleSuccessfulResponse(it, onSuccess, onError) },
-        { onError(it) }
+        attributes,
+        { handleSuccessfulResponse(it, cacheKey, true, onSuccess, onError) },
+        { handleErrorResponse(it, cacheKey, onSuccess, onError) }
     )
   }
 
+  private fun buildAttributesJson(
+      deviceId: String
+  ): JsonObject {
+    return JsonObject().apply {
+      add("visitor", JsonObject().apply {
+        addProperty("id", deviceId)
+      })
+    }
+  }
+
+  private fun buildCacheKey(
+      placementId: String,
+      placementMode: PlacementMode,
+      previewOptions: PlacementPreviewOptions,
+      attributes: JsonObject
+  ): String {
+    return StringBuilder().apply {
+      append(placementId)
+      append(placementMode.name)
+      append(previewOptions.campaignId)
+      append(previewOptions.experienceId)
+      append(attributes.toString())
+    }.toString()
+  }
+
   private fun getPlacementApiHost(configurationRepository: ConfigurationRepository): String {
-    return configurationRepository.load().placementApiHost
+    return configurationRepository.load()?.placementApiHost
         ?: ConfigurationModel.getDefault().placementApiHost
   }
 
   private fun handleSuccessfulResponse(
```

Comment:
I think methods handleSuccessfulReponse and handleErrorResponse are complex now. Especially when handleErrorResponse call handleSuccessfulReponse.
I would try to extract function which generates placement from placementModel.